plugins/obs-filters/drop-shadow-filter.c

```diff
@@ -0,0 +1,418 @@
+		struct drop_shadow_filter_data *filter,
+		gs_texrender_t *shadow_target, gs_texrender_t *shadow_source,
+		float kawase_distance)
+{
+	gs_texrender_reset(shadow_target);
+	if (gs_texrender_begin(shadow_target,
+			filter->total_width, filter->total_height)) {
+
+		gs_ortho(0, (float)filter->total_width, 0,
+				(float)filter->total_height, -1, 1);
+		struct vec4 black;
+		vec4_zero(&black);
+		gs_clear(GS_CLEAR_COLOR | GS_CLEAR_DEPTH, &black, 0, 0);
+
+		set_parameters(filter, shadow_source, kawase_distance);
+		set_image_parameter(filter);
+
+		gs_texture_t *output_texture = gs_texrender_get_texture(
+				shadow_target);
+		while (gs_effect_loop(filter->effect, "BlurPass")) {
+			gs_draw_sprite(output_texture, 0,
+					filter->total_width,
+					filter->total_height);
+		}
+
+		gs_texrender_end(shadow_target);
+	}
+}
+
+typedef struct {
+	int passes;
+	float distances[7];
+} kawase_entry_t;
+
+kawase_entry_t kawase_table[] = {
+	{2, {0.0, 0.0}},
+	{2, {0.0, 0.5}},
+	{2, {0.5, 0.5}},
+	{2, {0.5, 1.5}},
+	{3, {0.5, 1.5, 1.5}},
+	{3, {0.5, 1.5, 2.5}},
+	{4, {0.5, 0.5, 1.5, 2.5}},
+	{4, {0.5, 1.5, 1.5, 2.5}},
+	{4, {0.5, 1.5, 2.5, 2.5}},
+	{5, {0.5, 0.5, 1.5, 2.5, 3.5}},
+	{5, {0.5, 1.5, 1.5, 2.5, 3.5}},
+	{5, {0.5, 1.5, 2.5, 2.5, 3.5}},
+	{5, {0.5, 1.5, 2.5, 3.5, 3.5}},
+	{5, {0.5, 1.5, 2.5, 3.5, 4.5}},
+	{6, {0.5, 0.5, 1.5, 2.5, 3.5, 4.5}},
+	{6, {0.5, 1.5, 1.5, 2.5, 3.5, 4.5}},
+	{6, {0.5, 1.5, 2.5, 2.5, 3.5, 4.5}},
+	{6, {0.5, 1.5, 2.5, 3.5, 3.5, 4.5}},
+	{6, {0.5, 1.5, 2.5, 3.5, 4.5, 4.5}},
+	{6, {0.5, 1.5, 2.5, 3.5, 4.5, 5.5}},
+	{7, {0.5, 1.5, 1.5, 2.5, 3.5, 4.5, 5.5}},
+	{7, {0.5, 1.5, 2.5, 2.5, 3.5, 4.5, 5.5}},
+	{7, {0.5, 1.5, 2.5, 3.5, 3.5, 4.5, 5.5}},
+	{7, {0.5, 1.5, 2.5, 3.5, 4.5, 4.5, 5.5}},
+	{7, {0.5, 1.5, 2.5, 3.5, 4.5, 5.5, 5.5}},
+	{7, {0.5, 1.5, 2.5, 3.5, 4.5, 5.5, 6.5}}
+};
+
+static kawase_entry_t get_kawase_entry(int blur_radius)
+{
+	if (blur_radius > 25) {
+		blur_radius = 25;
+	}
+
+	return kawase_table[blur_radius];
+}
+
+static void drop_shadow_render(void *data, gs_effect_t *effect)
+{
+	struct drop_shadow_filter_data *filter = data;
+
+	gs_texrender_t *front = filter->shadow1_target;
+	gs_texrender_t *back = filter->shadow2_target;
+
+	kawase_entry_t kawase_entry = get_kawase_entry(filter->blur_radius);
+
+	// Start the rendering of the filter, first by processing
+	// the rest of the filter chain, filter->context->filter_texrender
+	// will now contain the result of the chain.
+	if (obs_source_process_filter_begin(filter->context, GS_RGBA,
+			OBS_ALLOW_DIRECT_RENDERING)) {
+
+		int pass = 0;
+		drop_shadow_render_first_pass(filter, back, front,
+			kawase_entry.distances[pass]);
+		SWAP(gs_texrender_t *, front, back);
+
+		for (; pass < (kawase_entry.passes - 1); pass++) {
+			drop_shadow_render_blur_pass(filter, back, front,
+					kawase_entry.distances[pass]);
+			SWAP(gs_texrender_t *, front, back);
+		}
+
+		// Last pass will blur one more time and composit the original
```

Comment:
This line and the next seem to be duplicates?

Style:  Multi-line comments should use `/* */`.